### ISSUE TITLE
KotlinProbeInstrumentation does precisely what gRPC Javadoc says not to

### ISSUE THREAD

**Issue by asarkar** [open-telemetry/opentelemetry-java-instrumentation#1411]
[CoroutineContextWrapper.tracingSuspend()](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/master/instrumentation/kotlin-coroutines/src/main/java/io/opentelemetry/javaagent/instrumentation/kotlincoroutines/KotlinProbeInstrumentation.java#L194):
```
public void tracingSuspend() {
    myTracingContext = Context.current();
    myTracingContext.detach(prevTracingContext);
}
```
Apart from saving the current context, the above code is equivalent to `Context.current().detach()`, which gRPC [Context.detach()](https://grpc.github.io/grpc-java/javadoc/io/grpc/Context.html#detach-io.grpc.Context-) asks not to do:
> **Never** use `Context.current().detach()`, as this will compromise this error-detecting mechanism.

On an aside, I noticed that gRPC context is used everywhere for propagation. This is very odd; OTEL ought to make their own. Most people wouldn't expect or want a tracing library to quietly pull in gRPC and its dependencies (Guava etc.).

**Comment by anuraaga** [comment]
Thanks for reporting - I'm currently working on upgrading our sdk dependency to remove the grpc dep.

https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/1384

I also noticed the exact same issue about kotlon as you did while doing so. This context current detach seems strange to me but currently our unit test fails without it for some reason. I've left a TODO for now to explore more why that could be happening.

**Comment by asarkar** [comment]
@anuraaga I'm playing around with the Kotlin instrumentation, if I find something useful, I'll post here. Can you provide a link to the unit test so that I can see what it's doing?

**Comment by anuraaga** [comment]
Thanks - the stress test "test concurrent suspend functions" fails if detaching myContext.

https://www.github.com/open-telemetry/opentelemetry-java-instrumentation/tree/master/instrumentation%2Fkotlin-coroutines%2Fsrc%2Ftest%2Fgroovy%2FKotlinCoroutineInstrumentationTest.groovy

**Comment by asarkar** [comment]
@anuraaga It seems the current implementation is completely detached from fundamental design of coroutine scope/context. I think the trace should be propagated as a `CoroutineContext.Element`, like it is shown in https://github.com/openzipkin/brave/issues/820#issuecomment-447614394. In that regard, it should be similar to the `Job` hierarchy; see [Coroutine Context and Scope](https://medium.com/@elizarov/coroutine-context-and-scope-c8b255d59055).

In layman's terms, for every new coroutine launched:
1. If a trace exists in the parent context, continue it; if not, create a new one.
2. Create a new span.
3. Attach closing of the new span with the current `Job`.

The coroutine context will manage the suspension/resumption without any help from OTEL.

**Comment by anuraaga** [comment]
The idea of using the element came up here

https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/1168#discussion_r483427587

I didn't have a strong enough reason to push forward on exploring it, but maybe we do want to reexamine this @johnbley.

**Comment by asarkar** [comment]
I think having to use a hokey gRPC context should be strong enough indication to look for a more “natural” mechanism for coroutines.

**Comment by johnbley** [comment]
I'd certainly welcome a cleaner way to do this that passes the test - I tried several times/ways and was unable to get it to work.  As for the Context oddity - yes, unfortunately with the visibility points in question here the instrumentation has no idea what spans/contexts will be created during the course of the next continuation, so it needs to capture the context "as it is now" instead of assuming that the context it set on `resume` will be the same as at the one at the next `suspend` - basically the opposite of a lot of our synch (and async) instrumentation.

**Comment by asarkar** [comment]
@johnbley 
> instead of assuming that the context it set on resume will be the same as at the one at the next suspend

I don't follow. It's the coroutine library's job to restore the context, just like it's the OS' job when threads are used. As long as OTEL relies on the context for propagation, where's the problem? The Brave code I linked to earlier does exactly that.

**Comment by trask** [comment]
@asarkar would you mind submitting a PR for this? It sounds like there's agreement that this could be improved, and we just need a concrete proposal that passes the tests.

**Comment by asarkar** [comment]
@trask It’s not yet clear to me how the proposal I made [here](https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/1411#issuecomment-711456437) would be implemented. Note that I’m interested in a library module that can be used without a javaagent. I tried an experiment locally but the [intercepts didn’t work](https://stackoverflow.com/q/64414060/839733).
